contributors/design-proposals/service-injection-policy.md

```diff
@@ -0,0 +1,274 @@
+management.
+
+The resultant modified spec will be annotated to show that it was modified by
+the `ServiceInjectionPolicy`.
+
+#### Optional Behavior with implicit injections
+
+The above described the behavior for explicitly defining what should be
+injected into a pod.
+
+But a helper has been discussed to automatically fill in the
+`ServiceInjectionSpec` values based off annotations on a service defining the
+resources you need. This would be similar to the functionality ConfigMaps and
+Secrets.
+
+## Examples
+
+### Simple Pod Spec Example
+
+User submitted pod spec:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: website
+  labels:
+    app: website
+    role: frontend
+spec:
+  containers:
+    - name: website
+      image: ecorp/website
+      ports:
+        - containerPort: 80
+```
+
+Example service injection policy:
+
+```yaml
+kind: ServiceInjectionPolicy
+apiVersion: extensions/v1beta1
+metadata:
+  name: allow-database
+  namespace: myns
+spec:
+  service: db
+  selector:
+    matchLabels:
+      role: frontend
+  env:
+    - name: DB_PORT
+      value: 6379
+  volumeMounts:
+    - mountPath: /cache
+      name: cache-volume
+  volumes:
+    - name: cache-volume
+      emptyDir: {}
+```
+
+Pod spec after admission controller:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: website
+  labels:
+    app: website
+    role: frontend
+  annotations:
+    serviceinjectionpolicy.k8s.io/db: allow-database
+spec:
+  containers:
+    - name: website
+      image: ecorp/website
+      volumeMounts:
+        - mountPath: /cache
+          name: cache-volume
+      volumes:
+        - name: cache-volume
+          emptyDir: {}
+      ports:
+        - containerPort: 80
+      env:
+        - name: DB_PORT
+          value: 6379
+```
+
+### ReplicaSet with Pod Spec Example
+
+User submitted ReplicaSet:
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: ReplicaSet
+metadata:
+  name: frontend
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      tier: frontend
+    matchExpressions:
+      - {key: tier, operator: In, values: [frontend]}
+  template:
+    metadata:
+      labels:
+        app: guestbook
+        tier: frontend
+    spec:
+      containers:
+      - name: php-redis
+        image: gcr.io/google_samples/gb-frontend:v3
+        resources:
+          requests:
+            cpu: 100m
+            memory: 100Mi
+        env:
+          - name: GET_HOSTS_FROM
+            value: dns
+        ports:
+          - containerPort: 80
+```
+
+Example service injection policy:
+
+```yaml
+kind: ServiceInjectionPolicy
+apiVersion: extensions/v1beta1
+metadata:
+  name: allow-database
+  namespace: myns
+spec:
+  service: db
+  selector:
+    matchLabels:
+      tier: frontend
+  env:
+    - name: DB_PORT
+      value: 6379
+  volumeMounts:
+    - mountPath: /cache
+      name: cache-volume
+  volumes:
+    - name: cache-volume
+      emptyDir: {}
+```
+
+ReplicaSet after admission controller modified pod spec:
```

Comment:
This should be a pod created from the replica set, rather than the replica set itself; non-pod resources won't be mutated.